### ISSUE TITLE
[BUG] Network issues on `2.24.6`

### ISSUE THREAD

**Issue by xtaixe** [docker/compose#11510]
### Description

We are seeing test failures due to `UnknownHostException` and `The connection attempt failed` after upgrading from `2.24.5` to `2.24.6` without any other changes.

This is happening both locally and on CI. The Docker Environment provided below is from Mac laptop with Docker Desktop where I installed version `2.24.6` manually to reproduce the issue which was detected in CI.

I don't have a reproducer I can share for now. Shooting in the dark here, but wondering if this could be related to #11429, since it seems to be the only change related to networking?

### Compose Version

```Text
v2.24.6
```

### Docker Environment

```Text
Client:
 Version:    25.0.2
 Context:    desktop-linux
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.12.1-desktop.4
    Path:     /Users/xtaixe/.docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.24.6
    Path:     /Users/xtaixe/.docker/cli-plugins/docker-compose
  debug: Get a shell into any image or container. (Docker Inc.)
    Version:  0.0.22
    Path:     /Users/xtaixe/.docker/cli-plugins/docker-debug
  dev: Docker Dev Environments (Docker Inc.)
    Version:  v0.1.0
    Path:     /Users/xtaixe/.docker/cli-plugins/docker-dev
  extension: Manages Docker extensions (Docker Inc.)
    Version:  v0.2.21
    Path:     /Users/xtaixe/.docker/cli-plugins/docker-extension
  feedback: Provide feedback, right in your terminal! (Docker Inc.)
    Version:  v1.0.4
    Path:     /Users/xtaixe/.docker/cli-plugins/docker-feedback
  init: Creates Docker-related starter files for your project (Docker Inc.)
    Version:  v1.0.0
    Path:     /Users/xtaixe/.docker/cli-plugins/docker-init
  sbom: View the packaged-based Software Bill Of Materials (SBOM) for an image (Anchore Inc.)
    Version:  0.6.0
    Path:     /Users/xtaixe/.docker/cli-plugins/docker-sbom
  scout: Docker Scout (Docker Inc.)
    Version:  v1.3.0
    Path:     /Users/xtaixe/.docker/cli-plugins/docker-scout
WARNING: Plugin "/usr/local/lib/docker/cli-plugins/docker-compose-old" is not valid: plugin candidate "compose-old" did not match "^[a-z][a-z0-9]*$"
WARNING: Plugin "/Users/xtaixe/.docker/cli-plugins/docker-compose.old" is not valid: plugin candidate "compose.old" did not match "^[a-z][a-z0-9]*$"
WARNING: Plugin "/Users/xtaixe/.docker/cli-plugins/docker-scan" is not valid: failed to fetch metadata: fork/exec /Users/xtaixe/.docker/cli-plugins/docker-scan: no such file or directory

Server:
 Containers: 0
  Running: 0
  Paused: 0
  Stopped: 0
 Images: 51
 Server Version: 25.0.2
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: ae07eda36dd25f8a1b98dfbf587313b99c0190bb
 runc version: v1.1.12-0-g51d5e94
 init version: de40ad0
 Security Options:
  seccomp
   Profile: unconfined
  cgroupns
 Kernel Version: 6.6.12-linuxkit
 Operating System: Docker Desktop
 OSType: linux
 Architecture: aarch64
 CPUs: 10
 Total Memory: 7.755GiB
 Name: docker-desktop
 ID: 0e3058a9-1f89-475a-a508-6d8143e2b96a
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 HTTP Proxy: http.docker.internal:3128
 HTTPS Proxy: http.docker.internal:3128
 No Proxy: hubproxy.docker.internal
 Experimental: false
 Insecure Registries:
  hubproxy.docker.internal:5555
  127.0.0.0/8
 Live Restore Enabled: false

WARNING: daemon is not using the default seccomp profile
```

**Comment by glours** [comment]
Hello @xtaixe 
Is it possible for you to bump the version of the engine on your CI (it will be more complicated to update it on Mac) to version `25.0.3`? There have been a couple of fixes around networking in this release

**Comment by xtaixe** [comment]
@glours unfortunately, is not possible.

**Comment by bf** [comment]
beautiful, stuck on same issue since two days :) 
only 3 out of 4 user-defined networks are created. looks like race condition.
docker compose 2.24.6 on 6.7.5-arch1-1

**Comment by bf** [comment]
@xtaixe I would install 25.0.3 but it is not available on archlinux extra-testing repo

**Comment by bf** [comment]
@glours so my machine now has **1022 network interfaces**! It might be that docker-compose creates too many interfaces and hits a hard limit in the linux kernel or NetworkManager, and then it fails to create further networks!

At the same time, docker-compose down does not properly delete created networks. Also `docker network prune` does not remove them. I think this is an issue to be looked at!

```
$ ip a
[...]
1020: veth4cdfcc8@if1019: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master br-6ec229210072 state UP group default 
    link/ether 5e:0e:5d:7b:c8:22 brd ff:ff:ff:ff:ff:ff link-netnsid 1
    inet6 fe80::5c0e:5dff:fe7b:c822/64 scope link proto kernel_ll 
       valid_lft forever preferred_lft forever
1022: veth72238e3@if1021: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master br-faf8ecb4c76e state UP group default 
    link/ether e2:97:9c:b7:6a:04 brd ff:ff:ff:ff:ff:ff link-netnsid 2
    inet6 fe80::e097:9cff:feb7:6a04/64 scope link proto kernel_ll 
       valid_lft forever preferred_lft forever
```

**Comment by ybc37** [comment]
@bf On Arch you can downgrade the package if it's still in the local cache (if not in your cache, see [Arch Linux Archive](https://wiki.archlinux.org/title/Arch_Linux_Archive)):

```sh
sudo pacman -U /var/cache/pacman/pkg/docker-compose-2.24.5-1-x86_64.pkg.tar.zst
```

When upgrading the system, ignore the package until a fixed version is in the repository, e.g.:

```sh
sudo pacman -Syu --ignore='docker-compose'
```

But be aware that partial upgrades may cause trouble: https://wiki.archlinux.org/title/System_maintenance#Partial_upgrades_are_unsupported

Obviously, just a workaround.

**Comment by bf** [comment]
this "PrefAddress:invalid IP" bug seems to be known since oct 2023 in https://github.com/moby/moby/issues/46615 and a fix at https://github.com/moby/moby/pull/46630 is also not active. FML

**Comment by bf** [comment]
I have fixed my issue. TLDR: Somewhere in a Dockerfile a **VOLUME** command is used but not connected in the docker-compose.yml. When commenting out the line with VOLUME in the respective Dockerfile, the random arp / hostname resolving errors of the user-defined bridge networks suddenly disappeared.

It's extremely weird. So for some user-defined bridge network interfaces docker just forgets to set the hostname of the other container in the network. I have containers with 4 user-defined networks and 3 work fine, but 1 is connected but the counterparty is missing.

The Dockerfile with the VOLUME command does not belong to the container(s) where the network issues appeared. It's like a race condition where at some point docker just seems to stop processing the arp announcements within the user-defined bridge networks (or it just stops connecting a container to the 3rd network out of 4).

Docker then frantically tries to resolve the expected hostname via the host dns server which obviously fails, but that at least shows up very prominently in `journalctl -u docker -f`.

However, the `wait-for-it.sh` scripts are still bugging out from time to time when used in a container connected to three networks (e.g. postgresql database network, redis cache network and upstream network for communication with proxy). When I refactored `docker-compose.yml` to use `healthcheck` and `depends_on: myservice: condition: service_healthy` for startup then the `wait-for-it.sh` script in the container Dockerfile was not needed any more and the error is resolved.

I hope this helps someone, I aged a year.

**Comment by glours** [comment]
@bf @xtaixe @ybc37 we bumped the version of compose-go yesterday into the main branch of Compose with a fix about networks order & priority, can you check if the [latest binary of Compose ](https://github.com/docker/compose/actions/runs/7994137875/artifacts/1264074044) potentially fixes your issue?

**Comment by kelhelawy** [comment]
Hello,
I am in the same team as @xtaixe, I tested the binary, and with some quick tests that were failing with `2.24.6`, seems they are working now, giving in `2.24.6` tests were flaky, would be good to hear the feedback from others as well.

**Comment by gileri** [comment]
>can you check if the [latest binary of Compose ](https://github.com/docker/compose/actions/runs/7994137875/artifacts/1264074044)potentially fixes your issue

I can confirm that while I hit the bug with 2.24.6, I do not hit it with this binary. Thanks for the heads up @glours !

**Comment by glours** [comment]
Is it ok to close the issue?

**Comment by milas** [comment]
This fix has been included in Compose [2.24.7+](https://github.com/docker/compose/releases/latest). If you continue to have problems after upgrading, comment here and we can re-open the issue or create a new one as appropriate. Thanks for the report!
